lib/command.js

```diff
@@ -1249,6 +1233,88 @@ Expecting one of '${allowedValues.join("', '")}'`);
     return result;
   }
 
+  /**
+   * Call parseArgs with extra handling:
+   * - custom error message if parseArgs throws 'commander.invalidArgument'
+   * - if the previous value is a promise, chain the call to resolve the promise before parsing
+   *
+   * @param {Option | Argument} target
+   * @param {string} value
+   * @param {Promise<any> | any} previous
+   * @param {string} invalidArgumentMessage
+   * @api private
+   */
+
+  _parseArg(target, value, previous, invalidArgumentMessage) {
+    const refineError = (err) => {
+      if (err.code === 'commander.invalidArgument') {
+        const message = `${invalidArgumentMessage} ${err.message}`;
+        this.error(message, { exitCode: err.exitCode, code: err.code });
+      }
+      throw err;
+    };
+
+    let result;
+    if (thenable(previous)) {
+      result = previous.then(resolvedPrevious => target.parseArg(value, resolvedPrevious));
+    } else {
+      try {
+        result = target.parseArg(value, previous);
+      } catch (err) {
+        refineError(err);
+      }
+    }
+
+    if (thenable(result)) result.catch(refineError);
```

Comment:
`refineError()` throws, so there will be an unhandled promise rejection. The result of the `catch()` call needs to be assigned to `result` to prevent it.

But if `previous` is already thenable, it is better to call `catch()` in the `previous.then()` callback (and return the result). Otherwise, `catch()` might be called on an already failed promise chain in which an error has already been caught and rethrown (because that is what `refineError()` does), so `refineError()` would be called again which does not make sense.

Reply:
Also, I think calling `result.then(x => x, refineError)` instead of `result.catch(refineError)` is more robust.

Reply:
> The result of the `catch()` call needs to be assigned to `result`

Oops, yes, good catch.

> Otherwise, `catch()` might be called on an already failed promise chain in which an error has already been caught and rethrown

I think this is correct pattern, but I'll try it before explaining my reasoning.

> Also, I think calling result.then(x => x, refineError) instead of result.catch(refineError) is more robust.

Because we only test for `.then()`? Yes, I have thought about that in past and then forgot. I want to just assume they are promise-like since the primary use-case is for async, but it isn't hard to only rely on `then` and cover a few more edge uses.

Reply:
> But if `previous` is already thenable, it is better to call `catch()` in the `previous.then()` callback (and return the result). Otherwise, `catch()` might be called on an already failed promise chain in which an error has already been caught and rethrown (because that is what `refineError()` does), so `refineError()` would be called again which does not make sense.

Took me a while to understand this. I was expecting a long catch chain, but see it can be avoided. 👍